model_templates/training/r_lang/custom.R

```diff
@@ -0,0 +1,191 @@
+init <- function(...) {
+  # custom init function to load required libraries
+  library(tidyverse)
+  library(caret)
+  library(recipes)
+  library(e1071)
+  library(gbm)
+}
+
+fit <- function(X, y, output_dir, class_order=NULL, row_weights=NULL, ...){
+    #' User-provided fit method, required for custom training
+    #'
+    #' Trains a regression or classification model using gbm (via caret)
+    #' @param X data.frame - training data to perform fit on
+    #' @param y data.frame column or array - target data to perform fit on
+    #' @param output_dir the path to write output. This is the path provided in '--output' parameter of the
+    #' 'drum fit' command.
+    #' @param class_order : A two element long list dictating the order of classes which should be used for
+    #'  modeling. Class order will always be passed to fit by DataRobot for classification tasks,
+    #'  and never otherwise. When models predict, they output a likelihood of one class, with a
+    #'  value from 0 to 1. The likelihood of the other class is 1 - this likelihood. Class order
+    #'  dictates that the first element in the list will be the 0 class, and the second will be the
+    #'  1 class.
+    #' @param row_weights: An array of non-negative numeric values which can be used to dictate how important
+    #'  a row is. Row weights is only optionally used, and there will be no filtering for which
+    #'  custom models support this. There are two situations when values will be passed into
+    #'  row_weights, during smart downsampling and when weights are explicitly provided by the user
+    #' @param ...: Added for forwards compatibility
+    #' @return Nothing
+
+  # set up dataframe for modeling
+  train_df <- X
+  train_df$target <- unlist(y)
+
+  if (!is.null(class_order)){
+    # maybe add more processing to y if needed? R might not care
+    outfile <- 'r_classif.rds'
+  }else{
+    outfile <-'r_reg.rds'
+  }
+
+  # set up the modeling pipeline
+  model_recipe <- recipe(target ~ ., data = train_df) %>%
+    # Drop constant columns
+    step_zv(all_predictors()) %>%
+    # Numeric preprocessing
+    step_medianimpute(all_numeric()) %>%
+    step_normalize(all_numeric(), -all_outcomes()) %>%
+    # Categorical preprocessing
+    step_other(all_nominal(), -all_outcomes()) %>%
+    step_dummy(all_nominal(), -all_outcomes())
+
```

Comment:
would you mind following the pattern of the other training models, and move this into another file, just to show users how imports work?

Reply:
so far the only foolproof way I've found to do this involves moving the actual fitting step into the other file along with the "recipe" code- if that's cool, I'll push that change. otherwise I may have to confer with an R expert on best-practices